### ISSUE TITLE
可运行的极简GPT

### ISSUE THREAD

**Issue by qianglifeng** [qianglifeng/IssueShared#2]
import torch
import torch.nn as nn
from torch.nn import functional as F

# --- 1. 超參數 (Hyperparameters) ---
batch_size = 32        # 一批訓練數據的大小
block_size = 128       # 上下文長度 (context length)
max_iters = 5000       # 最大訓練迭代次數
eval_interval = 500    # 每隔多少次迭代，評估一次損失
learning_rate = 1e-3   # 學習率
device = 'cuda' if torch.cuda.is_available() else 'cpu' # 決定使用 GPU 還是 CPU
eval_iters = 200
n_embd = 256           # 詞嵌入維度
n_head = 4             # 注意力頭的數量
n_layer = 4            # Transformer 模塊的層數
dropout = 0.2          # Dropout 比例
# ------------------------------------

print(f"Using device: {device}")

# --- 2. 數據準備 (Data Preparation) ---
# 我們直接使用一小段莎士比亞的文本作為示例
with open('input.txt', 'w', encoding='utf-8') as f:
    f.write("""First Citizen:
Before we proceed any further, hear me speak.

All:
Speak, speak.

First Citizen:
You are all resolved rather to die than to famish?

All:
Resolved. resolved.

First Citizen:
First, you know Caius Marcius is chief enemy to the people.

All:
We know't, we know't.
""")

with open('input.txt', 'r', encoding='utf-8') as f:
    text = f.read()

# 創建字符級的詞彙表
chars = sorted(list(set(text)))
vocab_size = len(chars)
# 創建字符到整數的映射 (stoi) 和整數到字符的映射 (itos)
stoi = { ch:i for i,ch in enumerate(chars) }
itos = { i:ch for i,ch in enumerate(chars) }
encode = lambda s: [stoi[c] for c in s] # 編碼器: 將字符串轉換為整數列表
decode = lambda l: ''.join([itos[i] for i in l]) # 解碼器: 將整數列表轉換為字符串

# 將所有文本數據轉換為 torch.Tensor
data = torch.tensor(encode(text), dtype=torch.long)
# 分割訓練集和驗證集 (90% 訓練, 10% 驗證)
n = int(0.9*len(data))
train_data = data[:n]
val_data = data[n:]

# 數據加載器 (Data Loader)
def get_batch(split):
    data = train_data if split == 'train' else val_data
    # 隨機選取 batch_size 個起始位置
    ix = torch.randint(len(data) - block_size, (batch_size,))
    # 根據起始位置，構建輸入序列 x 和目標序列 y
    x = torch.stack([data[i:i+block_size] for i in ix])
    y = torch.stack([data[i+1:i+block_size+1] for i in ix])
    x, y = x.to(device), y.to(device)
    return x, y

@torch.no_grad()
def estimate_loss():
    """評估模型在訓練集和驗證集上的損失"""
    out = {}
    model.eval() # 將模型設置為評估模式
    for split in ['train', 'val']:
        losses = torch.zeros(eval_iters)
        for k in range(eval_iters):
            X, Y = get_batch(split)
            logits, loss = model(X, Y)
            losses[k] = loss.item()
        out[split] = losses.mean()
    model.train() # 將模型重新設置為訓練模式
    return out

# --- 3. 模型定義 (Model Definition) ---
# (這部分與之前的回答完全相同)
class Head(nn.Module):
    def __init__(self, head_size):
        super().__init__()
        self.key = nn.Linear(n_embd, head_size, bias=False)
        self.query = nn.Linear(n_embd, head_size, bias=False)
        self.value = nn.Linear(n_embd, head_size, bias=False)
        self.register_buffer('tril', torch.tril(torch.ones(block_size, block_size)))
        self.dropout = nn.Dropout(dropout)

    def forward(self, x):
        B, T, C = x.shape
        k = self.key(x)
        q = self.query(x)
        wei = q @ k.transpose(-2, -1) * C**-0.5
        wei = wei.masked_fill(self.tril[:T, :T] == 0, float('-inf'))
        wei = F.softmax(wei, dim=-1)
        wei = self.dropout(wei)
        v = self.value(x)
        out = wei @ v
        return out

class MultiHeadAttention(nn.Module):
    def __init__(self, num_heads, head_size):
        super().__init__()
        self.heads = nn.ModuleList([Head(head_size) for _ in range(num_heads)])
        self.proj = nn.Linear(n_embd, n_embd)
        self.dropout = nn.Dropout(dropout)

    def forward(self, x):
        out = torch.cat([h(x) for h in self.heads], dim=-1)
        out = self.dropout(self.proj(out))
        return out

class FeedFoward(nn.Module):
    def __init__(self, n_embd):
        super().__init__()
        self.net = nn.Sequential(
            nn.Linear(n_embd, 4 * n_embd),
            nn.ReLU(),
            nn.Linear(4 * n_embd, n_embd),
            nn.Dropout(dropout),
        )

    def forward(self, x):
        return self.net(x)

class Block(nn.Module):
    def __init__(self, n_embd, n_head):
        super().__init__()
        head_size = n_embd // n_head
        self.sa = MultiHeadAttention(n_head, head_size)
        self.ffwd = FeedFoward(n_embd)
        self.ln1 = nn.LayerNorm(n_embd)
        self.ln2 = nn.LayerNorm(n_embd)

    def forward(self, x):
        x = x + self.sa(self.ln1(x))
        x = x + self.ffwd(self.ln2(x))
        return x

class GPTLanguageModel(nn.Module):
    def __init__(self):
        super().__init__()
        self.token_embedding_table = nn.Embedding(vocab_size, n_embd)
        self.position_embedding_table = nn.Embedding(block_size, n_embd)
        self.blocks = nn.Sequential(*[Block(n_embd, n_head=n_head) for _ in range(n_layer)])
        self.ln_f = nn.LayerNorm(n_embd)
        self.lm_head = nn.Linear(n_embd, vocab_size)

    def forward(self, idx, targets=None):
        B, T = idx.shape
        tok_emb = self.token_embedding_table(idx)
        pos_emb = self.position_embedding_table(torch.arange(T, device=device))
        x = tok_emb + pos_emb
        x = self.blocks(x)
        x = self.ln_f(x)
        logits = self.lm_head(x)

        loss = None
        if targets is not None:
            B, T, C = logits.shape
            logits = logits.view(B*T, C)
            targets = targets.view(B*T)
            loss = F.cross_entropy(logits, targets)
        return logits, loss

    def generate(self, idx, max_new_tokens):
        for _ in range(max_new_tokens):
            idx_cond = idx[:, -block_size:]
            logits, loss = self(idx_cond)
            logits = logits[:, -1, :]
            probs = F.softmax(logits, dim=-1)
            idx_next = torch.multinomial(probs, num_samples=1)
            idx = torch.cat((idx, idx_next), dim=1)
        return idx

# --- 4. 創建模型和優化器 ---
model = GPTLanguageModel()
m = model.to(device)
optimizer = torch.optim.AdamW(model.parameters(), lr=learning_rate)

# --- 5. 訓練循環 (Training Loop) ---
for iter in range(max_iters):
    if iter % eval_interval == 0:
        losses = estimate_loss()
        print(f"step {iter}: train loss {losses['train']:.4f}, val loss {losses['val']:.4f}")

    xb, yb = get_batch('train')
    logits, loss = model(xb, yb)
    optimizer.zero_grad(set_to_none=True)
    loss.backward()
    optimizer.step()

# --- 6. 生成文本 (Generation) ---
print("\n--- 訓練完成，開始生成文本 ---\n")
# 創建一個初始上下文，例如一個換行符 (表示新段落的開始)
context = torch.zeros((1, 1), dtype=torch.long, device=device)
# 使用模型生成 500 個 token
generated_code = m.generate(context, max_new_tokens=500)[0].tolist()
print(decode(generated_code))
